sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala

```diff
@@ -746,13 +746,11 @@ case class Last(child: Expression) extends PartialAggregate with trees.UnaryNode
 case class LastFunction(expr: Expression, base: AggregateExpression) extends AggregateFunction {
   def this() = this(null, null) // Required for serialization.
 
-  var result: Any = null
+  var result: MutableLiteral = MutableLiteral(null, expr.dataType)
 
   override def update(input: InternalRow): Unit = {
-    result = input
+    result.value = expr.eval(input)
```

Comment:
@ggupta81 I see the problem. I think we only need to change this line to fix it. Basically, the input row can be a mutable row. So, we need to eagerly evaluate the expression instead of just pointing result to the input.